File: gandi-publish-ds.pl
#!/usr/bin/perl

use strict;
use XML::RPC;
use Data::Dumper;
use Net::DNS;

my $endpoint = $ENV{'ENDPOINT'} || die "no endpoint in environnement system\n";
my $apikey = $ENV{'APIKEY'} || die "no apikey in environnement system\n";

print "Connecting to Gandi API endpoint $endpoint\n";
my $api = XML::RPC->new($endpoint);
my $version = $api->call('version.info', $apikey);
print "  Found API version: $version->{'api_version'}\n";

my $domains = $api->call('domain.list', $apikey);
chomp(my @domain = `/knot/sbin/knotc conf-read zone.domain | awk '{print \$NF}'`);

for(@domain) {
    my $domain = $_;
    $domain =~ s/\.$//;

    chomp(my @input = `/knot/sbin/knotc zone-read $domain @ CDNSKEY | cut -d' ' -f2-`);

    for(@input) {
        chomp(my(undef, undef, undef, $flag, undef, $alg, $public_key) = split /\s+/, $_, 7);
        print "Domain: $domain\n";
        print "Flag: $flag\n";
        print "Alg: $alg\n";
        print "Key: $public_key\n";

        my $rr = new Net::DNS::RR($_);
        my $keytag = $rr->keytag;
        print "Keytag: $keytag\n";

        my $found_domain = 0;
        for(@$domains) {
            if($_->{'fqdn'} eq $domain) {
                print "Domain $domain located in Gandi account\nChecking existing DNSSEC keys...\n";
                $found_domain++;

                my $keys = $api->call('domain.dnssec.list', $apikey, $domain);
                my $found_key = 0;
                for(@$keys) {
                    if($_->{'keytag'} eq $keytag && $_->{'flags'} eq $flag && $_->{'public_key'} eq $public_key) {
                        print "  Found existing matching key. No upload required.\n";
                        $found_key++;
                    }
                }

                if($found_domain && !$found_key) {
                    print "Domain $domain found on account, key not found, so key needs uploading...\n";
                    if(scalar(@$keys) >= 4) {
                        print "  Gandi has 4 key limit. Currently found ".scalar(@$keys)." keys\n";
                        exit 99;
                    }
                    my $params = {
                        'flags' => $flag,
                        'algorithm' => $alg,
                        'public_key' => $public_key
                    };
                    my $key_add = $api->call('domain.dnssec.create', $apikey, $domain, $params);
                    my $sub = `/knot/sbin/knotc zone-ksk-submitted $domain`;
                    print Dumper($key_add)."\n";
                }
            }
        }
        die "Didn't find $domain in account. Exiting.\n" unless $found_domain;
    }
}

File: gandi-publish-ds.py
#!/usr/bin/python3

import os, sys, re
import subprocess
import requests
import struct, base64

endpoint = os.environ['ENDPOINT'] if 'ENDPOINT' in os.environ else sys.exit('no endpoint in environnement system')
apikey   = os.environ['APIKEY'] if 'APIKEY' in os.environ else sys.exit('no apikey in environnement system')

def urlget(index, apikey):
    response = requests.get(
        f'{endpoint}{index}',
        headers={'authorization': f'Apikey {apikey}'}
    )
    return response

def calc_keyid(flags, protocol, algorithm, dnskey):
    st = struct.pack('!HBB', int(flags), int(protocol), int(algorithm))
    st += base64.b64decode(dnskey)

    cnt = 0
    for idx in range(len(st)):
        s = struct.unpack('B', st[idx:idx+1])[0]
        if (idx % 2) == 0:
            cnt += s << 8
        else:
            cnt += s

    return ((cnt & 0xFFFF) + (cnt >> 16)) & 0xFFFF

print(f'Connecting to Gandi API endpoint {endpoint}')
try:
    arrdomains = urlget('domain/domains', apikey)
    arrdomains.raise_for_status()
    print('  Found API version: v5')
except requests.exceptions.HTTPError as e:
    sys.exit(e)

arrdomain = subprocess.run("/knot/sbin/knotc conf-read zone.domain | awk '{print $NF}'", shell=True, check=True, capture_output=True, text=True).stdout.strip('\n')

for domains in arrdomains.json():
    for domain in arrdomain.split('\n'):
        domain = re.sub(r'\.$', '', domain)

        if domains['fqdn'] == domain:
            print(f'Domain {domain} located in Gandi account\nChecking existing DNSSEC keys...')

            arrzone = subprocess.run(f"/knot/sbin/knotc zone-read {domain} @ CDNSKEY | cut -d' ' -f2-", shell=True, check=True, capture_output=True, text=True).stdout.strip('\n')

            for zone in arrzone.split('\n'):
                cdnkey = zone.rstrip('\n').split()
                print(f'  Domain: {domain}')
                print(f'  Flag: {cdnkey[3]}')
                print(f'  Alg: {cdnkey[5]}')
                print(f'  Key: {cdnkey[6]}')

                keytag = calc_keyid(cdnkey[3], cdnkey[4], cdnkey[5], cdnkey[6])
                print(f'  Key: {keytag}')

        else:
            print(f"Didn't find {domain} in account. Exiting.")
